src/cv/ipcv.cpp

```diff
@@ -1,8 +1,218 @@
+
+    return good_squares;
+
+    // TODO: Debug macros?
+    /*
+    std::cout << "Good squares: " << std::to_string(good_squares.size()) << ". Bad Squares: " << std::to_string(squares.size()) << "\n";
+    draw_squares(image, good_squares);
+
+    std::string opath = path_string + "squar_" + file_title;
+
+    if (output_path != "") {
+        imwrite(opath, image);
+        std::cout << "Image has been written to " << opath << "\n";
+    }
+    */
+}
+
+// Extracts a quadrangle out of an image that is supposed to represent a whiteboard.
+Mat get_whiteboard(Mat image, std::string output_dir)
+{
+    // Locate squares in an image most likely to be the whiteboard
+    vector<vector<Point>> good_squares = locate_quadrangles(image, output_dir);
+
+    // If we didn't find any good squares, then we just have to give up, because
+    // whatever squares we did find are either totally wrong or encapsulate
+    // the whole image anyway.
+    if (good_squares.size() == 0)
+    {
+        std::cout << "Found no good squares :(\n";
+        return image;
+    }
+
+    // Get the biggest square 
+    vector<Point> best_square = good_squares[good_squares.size() - 1];
+
+    // Sort the corners 
+    sort_points_clockwise(best_square);
+
+    // Compute the bounding box of the contour
+    cv::Rect boundingBox = cv::boundingRect(best_square);
+
+    // Clockwise order
+    std::vector<cv::Point2f> dstPoints = {
+        {0, 0},
+        {(float) boundingBox.width - 1, 0},
+        {(float) boundingBox.width - 1, (float) boundingBox.height - 1},
+        {0, (float) boundingBox.height - 1},
+    };
+
+    Mat homography = findHomography(best_square, dstPoints, RANSAC);
+    
+    // Warp the perspective
+    cv::Mat warpedImage;
+    cv::warpPerspective(image, warpedImage, homography, cv::Size(boundingBox.width, boundingBox.height));
+
+    #ifdef INKPATH_DEBUG
+    if (output_dir != "") {
+        std::string opath = output_dir + "warped.jpg";
+        imwrite(opath, warpedImage);
+        std::cout << "Image has been written to " << opath << "\n";
+    }
+    #endif
+
+    return warpedImage;
+}
+
+// helper function: finds a cosine of angle between vectors from 
+// pt0->pt1 and from pt0->pt2
+double angle( Point pt1, Point pt2, Point pt0 )
+{
+    double dx1 = pt1.x - pt0.x;
+    double dy1 = pt1.y - pt0.y;
+    double dx2 = pt2.x - pt0.x;
+    double dy2 = pt2.y - pt0.y;
+    return (dx1*dx2 + dy1*dy2)/sqrt((dx1*dx1 + dy1*dy1)*(dx2*dx2 + dy2*dy2) + 1e-10);
+}
+
+// https://stackoverflow.com/a/8863060/6095682
+void find_squares(Mat& image, vector<vector<Point> >& squares)
+{
+    // Make a border around the whole image to help with detecting boards who go
+    // to the edge of the image
+    int border_width = 10;
+    int borderType = BORDER_CONSTANT;
+    Scalar value(0, 0, 0);
+    copyMakeBorder(image, image, border_width, border_width, border_width, border_width, borderType, value );
+
+    // blur will enhance edge detection
```

Comment:
I'm not sure about this. I had a bug where IMG_309.jpg wasn't detecting properly, and it was because the image was double-blurred.